src/vs/editor/contrib/dnd/browser/dnd.ts

```diff
@@ -0,0 +1,97 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import 'vs/css!./dnd';
+import { IDisposable, dispose } from 'vs/base/common/lifecycle';
+import { ICodeEditor, IEditorMouseEvent } from 'vs/editor/browser/editorBrowser';
+import { editorContribution } from 'vs/editor/browser/editorBrowserExtensions';
+import * as editorCommon from 'vs/editor/common/editorCommon';
+import { Position } from 'vs/editor/common/core/position';
+import { Selection } from 'vs/editor/common/core/selection';
+import { DragTargetHintWidget } from './dndHintWidget';
+import { DragAndDropCommand } from '../common/dragAndDropCommand';
+
+@editorContribution
+export class DragAndDropController implements editorCommon.IEditorContribution {
+
+	private static ID = 'editor.contrib.dragAndDrop';
+
+	private _editor: ICodeEditor;
+	private _toUnhook: IDisposable[];
+	private _targetWidget: DragTargetHintWidget;
+	private _active: boolean;
+	private _dragSelection: Selection;
+
+	static get(editor: editorCommon.ICommonCodeEditor): DragAndDropController {
+		return editor.getContribution<DragAndDropController>(DragAndDropController.ID);
+	}
+
+	constructor(editor: ICodeEditor) {
+		this._editor = editor;
+		this._toUnhook = [];
+		this._toUnhook.push(this._editor.onMouseDrag((e: IEditorMouseEvent) => this._onEditorMouseDrag(e)));
+		this._toUnhook.push(this._editor.onMouseDrop((e: IEditorMouseEvent) => this._onEditorMouseDrop(e)));
+		this._targetWidget = new DragTargetHintWidget(editor);
+		this._active = false;
+	}
+
+	private _onEditorMouseDrag(mouseEvent: IEditorMouseEvent): void {
+		let target = mouseEvent.target;
+
+		if (this._active) {
+			this._targetWidget.showAt(target.position);
+		} else {
+			let possibleSelections = this._editor.getSelections().filter(selection => selection.containsPosition(target.position));
+
+			if (possibleSelections.length === 1) {
+				this._active = true;
+				this._dragSelection = possibleSelections[0];
+				this._targetWidget.showAt(target.position);
+			}
+		}
+	}
+
+	private _onEditorMouseDrop(mouseEvent: IEditorMouseEvent): void {
+		let targetPosition = mouseEvent.target.position;
```

Comment:
addressed. Currently only support drop into CONTENT and CONTENT_EMPTY in current editor, except content widget.